File: highspeed_report/highspeed_accounts/report/simplified_product_profitability/simplified_product_profitability.py
# Copyright (c) 2026, Highspeed and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt

def execute(filters=None):
    if not filters:
        filters = {}

    columns = get_columns()
    data = get_data(filters)
    summary = get_report_summary(data)

    return columns, data, None, None, summary

def get_columns():
    return [
        {
            "label": _("Item Code"),
            "fieldname": "item_code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 140
        },
        {
            "label": _("Item Name"),
            "fieldname": "item_name",
            "fieldtype": "Data",
            "width": 180
        },
        {
            "label": _("Item Group"),
            "fieldname": "item_group",
            "fieldtype": "Link",
            "options": "Item Group",
            "width": 130
        },
        {
            "label": _("UOM"),
            "fieldname": "uom",
            "fieldtype": "Link",
            "options": "UOM",
            "width": 90
        },
        {
            "label": _("Sold Quantity"),
            "fieldname": "qty",
            "fieldtype": "Float",
            "width": 110
        },
        {
            "label": _("Sales Amount"),
            "fieldname": "sales_amount",
            "fieldtype": "Currency",
            "width": 130
        },
        {
            "label": _("Cost Amount"),
            "fieldname": "cost_amount",
            "fieldtype": "Currency",
            "width": 130
        },
        {
            "label": _("Gross Profit"),
            "fieldname": "gross_profit",
            "fieldtype": "Currency",
            "width": 130
        },
        {
            "label": _("Profit Margin %"),
            "fieldname": "margin_percentage",
            "fieldtype": "Percent",
            "width": 110
        }
    ]

def get_data(filters):
    conditions = []
    values = {
        "from_date": filters.get("from_date"),
        "to_date": filters.get("to_date")
    }

    if filters.get("company"):
        conditions.append("si.company = %(company)s")
        values["company"] = filters.get("company")
    if filters.get("item_group"):
        conditions.append("sii.item_group = %(item_group)s")
        values["item_group"] = filters.get("item_group")
    if filters.get("warehouse"):
        conditions.append("sii.warehouse = %(warehouse)s")
        values["warehouse"] = filters.get("warehouse")

    condition_str = " AND " + " AND ".join(conditions) if conditions else ""

    query = f"""
        SELECT 
            sii.item_code,
            sii.item_name,
            sii.item_group,
            sii.stock_uom as uom,
            SUM(sii.qty) as qty,
            SUM(sii.base_net_amount) as sales_amount,
            SUM(sii.qty * sii.incoming_rate) as cost_amount
        FROM 
            `tabSales Invoice Item` sii
        INNER JOIN 
            `tabSales Invoice` si ON sii.parent = si.name
        WHERE 
            si.docstatus = 1
            AND si.posting_date BETWEEN %(from_date)s AND %(to_date)s
            {condition_str}
        GROUP BY 
            sii.item_code, sii.item_name, sii.item_group, sii.stock_uom
        ORDER BY 
            sales_amount DESC
    """
    
    raw_data = frappe.db.sql(query, values, as_dict=True)
    
    processed_data = []
    for row in raw_data:
        sales = flt(row.sales_amount)
        cost = flt(row.cost_amount)
        profit = sales - cost
        margin = (profit / sales * 100) if sales else 0.0
        
        processed_data.append({
            "item_code": row.item_code,
            "item_name": row.item_name,
            "item_group": row.item_group,
            "uom": row.uom,
            "qty": row.qty,
            "sales_amount": sales,
            "cost_amount": cost,
            "gross_profit": profit,
            "margin_percentage": margin
        })
        
    return processed_data

def get_report_summary(data):
    total_sales = sum(row["sales_amount"] for row in data)
    total_cost = sum(row["cost_amount"] for row in data)
    total_profit = total_sales - total_cost
    avg_margin = (total_profit / total_sales * 100) if total_sales else 0.0

    return [
        {
            "value": total_sales,
            "label": _("Total Sales"),
            "indicator": "green",
            "datatype": "Currency"
        },
        {
            "value": total_cost,
            "label": _("Total Cost"),
            "indicator": "blue",
            "datatype": "Currency"
        },
        {
            "value": total_profit,
            "label": _("Gross Profit"),
            "indicator": "orange",
            "datatype": "Currency"
        },
        {
            "value": avg_margin,
            "label": _("Average Margin %"),
            "indicator": "purple",
            "datatype": "Percent"
        }
    ]

File: highspeed_report/highspeed_accounts/report/simplified_product_profitability/simplified_product_profitability.js
// Copyright (c) 2026, Highspeed and contributors
// For license information, please see license.txt

frappe.query_reports["Simplified Product Profitability"] = {
    "filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 1,
            "width": "200px"
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1,
            "width": "100px"
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1,
            "width": "100px"
        },
        {
            "fieldname": "item_group",
            "label": __("Item Group"),
            "fieldtype": "Link",
            "options": "Item Group",
            "width": "150px"
        },
        {
            "fieldname": "warehouse",
            "label": __("Warehouse"),
            "fieldtype": "Link",
            "options": "Warehouse",
            "get_query": function() {
                var company = frappe.query_report.get_filter_value('company');
                return {
                    "filters": {
                        "company": company,
                        "is_group": 0
                    }
                };
            },
            "width": "150px"
        }
    ],
    "formatter": function(value, row, column, data, default_formatter) {
        if (!data) {
            return default_formatter(value, row, column, data);
        }

        if (column.fieldname === "margin_percentage") {
            const margin = parseFloat(value) || 0;
            let val = default_formatter(value, row, column, data);
            
            if (margin < 5) {
                return `<span style="color: #e74c3c; font-weight: bold;">${val}</span>`;
            } else if (margin < 15) {
                return `<span style="color: #e67e22; font-weight: bold;">${val}</span>`;
            } else {
                return `<span style="color: #27ae60; font-weight: bold;">${val}</span>`;
            }
        }

        if (column.fieldname === "gross_profit") {
            const profit = parseFloat(value) || 0;
            let val = default_formatter(value, row, column, data);
            if (profit < 0) {
                return `<span style="color: #e74c3c; font-weight: bold;">${val}</span>`;
            } else if (profit > 0) {
                return `<span style="color: #27ae60; font-weight: bold;">${val}</span>`;
            }
        }

        if (data.is_total_row) {
            return `<span style="font-weight: bold; background-color: #f8fafc; padding: 2px;">${default_formatter(value, row, column, data)}</span>`;
        }

        return default_formatter(value, row, column, data);
    },
    "onload": function(report) {
        report.page.add_inner_button(__("طباعة التقرير"), function() {
            if (report.data && report.data.length > 0) {
                const html = get_profitability_print_html(report.data, report.get_values());
                var w = window.open();
                if (w) {
                    w.document.write(html);
                    w.document.close();
                    setTimeout(function() {
                        w.print();
                    }, 500);
                } else {
                    frappe.msgprint(__("تم حظر النوافذ المنبثقة. يرجى السماح بالنوافذ المنبثقة لطباعة التقرير."));
                }
            } else {
                frappe.msgprint(__("لا توجد بيانات لعرضها."));
            }
        });
    }
};

function get_profitability_print_html(data, filters) {
    let rows_html = '';
    let total_row = null;
    
    data.forEach(row => {
        if (row.is_total_row) {
            total_row = row;
            return;
        }
        rows_html += `
            <tr>
                <td>${row.item_code || ""}</td>
                <td style="text-align: right;">${row.item_name || ""}</td>
                <td>${row.item_group || ""}</td>
                <td>${row.stock_uom || ""}</td>
                <td class="en-number">${row.qty || 0}</td>
                <td class="en-number">${frappe.format(row.sales_amount, {fieldtype: 'Currency'})}</td>
                <td class="en-number">${frappe.format(row.cogs, {fieldtype: 'Currency'})}</td>
                <td class="en-number">${frappe.format(row.gross_profit, {fieldtype: 'Currency'})}</td>
                <td class="en-number" style="color: ${row.margin_percentage < 5 ? '#e74c3c' : (row.margin_percentage < 15 ? '#e67e22' : '#27ae60')}; font-weight: bold;">${(row.margin_percentage || 0).toFixed(2)}%</td>
            </tr>
        `;
    });
    
    if (total_row) {
        rows_html += `
            <tr style="font-weight: bold; background-color: #f2f2f2;">
                <td colspan="4" style="text-align: center;">${__("Total / الإجمالي")}</td>
                <td class="en-number">${total_row.qty || 0}</td>
                <td class="en-number">${frappe.format(total_row.sales_amount, {fieldtype: 'Currency'})}</td>
                <td class="en-number">${frappe.format(total_row.cogs, {fieldtype: 'Currency'})}</td>
                <td class="en-number">${frappe.format(total_row.gross_profit, {fieldtype: 'Currency'})}</td>
                <td class="en-number">${(total_row.margin_percentage || 0).toFixed(2)}%</td>
            </tr>
        `;
    }
    
    const currentDate = frappe.datetime.get_today();
    return `
    <!DOCTYPE html>
    <html dir="rtl">
    <head>
        <meta charset="UTF-8">
        <title>تقرير أرباح المنتجات</title>
        <style>
            @page { size: A4 landscape; margin: 0.5cm; }
            body { font-family: Arial, sans-serif; margin: 0; padding: 10px; direction: rtl; font-size: 11px; }
            .header-table { width: 100%; border-collapse: collapse; margin-bottom: 15px; }
            .header-table td { border: none; text-align: center; }
            .header-table h3 { margin: 5px 0; font-size: 16px; }
            .header-table h4 { margin: 5px 0; font-size: 14px; color: #555; }
            .info-table, .data-table { width: 100%; border-collapse: collapse; margin-bottom: 10px; }
            .info-table td, .info-table th { border: 1px solid #ddd; padding: 5px; text-align: right; }
            .data-table td, .data-table th { border: 1px solid #ccc; padding: 6px 4px; text-align: center; }
            .data-table th { background-color: #f5f5f5; font-weight: bold; }
            .data-table td { font-size: 10px; }
            .en-number { font-family: Arial, sans-serif !important; }
        </style>
    </head>
    <body>
        <div class="header-table">
            <table style="width: 100%; border: none;">
                <tr>
                    <td style="text-align: center;">
                        <h3>${filters.company || ""}</h3>
                        <h4>تقرير أرباح المنتجات المبسط / Product Profitability Report</h4>
                    </td>
                </tr>
            </table>
        </div>
        <table class="info-table">
            <tr>
                <td style="font-weight: bold; width: 15%;">الفترة / Period:</td>
                <td class="en-number" style="text-align: right;">${filters.from_date || ""} إلى ${filters.to_date || ""}</td>
                <td style="font-weight: bold; width: 15%;">تاريخ الطباعة:</td>
                <td class="en-number" style="text-align: right;">${currentDate}</td>
            </tr>
        </table>
        <table class="data-table">
            <thead>
                <tr>
                    <th>كود الصنف / Code</th>
                    <th>اسم الصنف / Name</th>
                    <th>المجموعة / Group</th>
                    <th>الوحدة / UOM</th>
                    <th>الكمية / Qty</th>
                    <th>المبيعات / Sales</th>
                    <th>التكلفة / Cost</th>
                    <th>الربح / Profit</th>
                    <th>الهامش / Margin</th>
                </tr>
            </thead>
            <tbody>
                ${rows_html}
            </tbody>
        </table>
    </body>
    </html>
    `;
}
